### PR TITLE
Testing harness for HTTP request parsing

### LINKED ISSUE #4: Create a harness for HTTP request parsing tests
HTTP request parsing involves handling:
- headers
- form values
- query parameters

Incorrect or non-deterministic handling of these might lead to security issues such as [request smuggling](https://portswigger.net/web-security/request-smuggling).

_[The Tangled Web](https://lcamtuf.coredump.cx/tangled/)_ by Michał Zalewski (lcamtuf) (2011), specifically chapters _I.2 It Starts with a URL_ and _I.3 Hypertext Transfer Protocol_, is a good resource for learning how complex and full of edge cases this task can be.

We need to create a testing harness where, as unit tests, we can send HTTP requests (built as arbitrary byte slices to cover attacks leveraging tricky encoding) and assert on what the `http.Handler` sees inside `http.Request`.

The assertions *must not* assume the client is written in Go.

**This issue considers only building the needed infrastructure, not the test scenarios themselves. These are to be tracked in separate issues.**

### PR DESCRIPTION
**Pull request by mattiasgrenfeldt**
Co-authored-by: Mattias Grenfeldt <grenfeldt@google.com>
Co-authored-by: Mara Mihali <maramihali@google.com>

Fixes #4 

We created the test harness for HTTP request parsing.

### CODE REVIEW

File: requestparsing/test_harness.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+type mockListener struct {
+	closeOnce      sync.Once
+	connChannel    chan net.Conn
+	serverEndpoint io.Closer
+	clientEndpoint net.Conn
+}
+
+// Creates a mock listener that passes requests to the HTTP server as part of
```
Comment (bsiegert) [inline requestparsing/test_harness.go:29]: Function docstrings should begin with the name of the function and end with punctuation.

// newMockListener creates a ... as part of the test harness.

Also: are these fakes instead of mocks?

File: requestparsing/test_harness.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		close(l.connChannel)
+	})
+	return err
+}
+
+func (l *mockListener) Addr() net.Addr {
+	return l.clientEndpoint.LocalAddr()
+}
+
+// SendRequest writes 'request' to the clientEndpoint connection which will //send the request to the server listening on this listener.
```
Comment (bsiegert) [inline requestparsing/test_harness.go:70]: nit:  linebreak in comment

File: requestparsing/test_harness.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	bytes := make([]byte, 4096)
+	n, err := l.clientEndpoint.Read(bytes)
+	if n == 4096 {
+		return nil, errors.New("response larger than 4096 bytes")
+	}
+	return bytes[:n], err
+}
+
+func makeRequest(ctx context.Context, req []byte, callbackFun func(*http.Request)) ([]byte, error) {
+
```
Comment (bsiegert) [inline requestparsing/test_harness.go:94]: nit: no empty line at the start of the function.

File: requestparsing/test_harness.go (hunk 1 above)
Comment (kele) [reply requestparsing/test_harness.go:29]: Docs: https://github.com/golang/go/wiki/CodeReviewComments#doc-comments

File: requestparsing/test_harness.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+)
+
+type fakeHandler struct {
```
Comment (kele) [inline requestparsing/test_harness.go:12]: Make this public.

File: requestparsing/test_harness.go (hunk 1 above)
Comment (mihalimara22) [reply requestparsing/test_harness.go]: I think they are fakes, renamed in [beb7c84](url) and refactored documentation.

File: requestparsing/test_harness.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+func (h *fakeHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.callback(r)
+	io.WriteString(w, "Hello world!")
+}
+
+type fakeListener struct {
+	closeOnce      sync.Once
+	connChannel    chan net.Conn
```
Comment (kele) [inline requestparsing/test_harness.go:25]: Avoid adding type information in the name. Consider simply `channels`.

File: requestparsing/test_harness.go (hunk 2 above)
Comment (mihalimara22) [reply requestparsing/test_harness.go]: Solved in  [beb7c84](url).

File: requestparsing/test_harness.go (hunk 3 above)
Comment (mihalimara22) [reply requestparsing/test_harness.go]: Solved in [beb7c84](url).

File: requestparsing/test_harness.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	ch, ok := <-l.connChannel
+	if !ok {
+		return nil, errors.New("Listener closed")
+	}
+	return ch, nil
+}
+
+// Close will close the two network connections and the listener
+func (l *fakeListener) Close() (err error) {
+	l.closeOnce.Do(func() {
```
Comment (kele) [inline requestparsing/test_harness.go:56]: This is a bit tricky.

On one hand, you are trying to use the named return value to save yourself creating a new variable and capture whatever is done in `Do()`, but you are also potentially leaving one endpoint open, even though it could've been closed.

Please find a way to refactor this.

File: requestparsing/test_harness.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	return &fakeListener{
+		connChannel:    c,
+		serverEndpoint: s2c,
+		clientEndpoint: c2s,
+	}
+}
+
+// Accept passes a network connection to the HTTP server to
+// enable bidirectional communication with the client. It will return an error
+// if Accept is called after the listener was closed
```
Comment (kele) [inline requestparsing/test_harness.go:45]: nit: here and every godoc comment: make them full sentences with correct punctuation (i.e. dot at the end).

File: requestparsing/test_harness.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+func (l *fakeListener) Addr() net.Addr {
+	return l.clientEndpoint.LocalAddr()
+}
+
+// SendRequest writes a request to the client endpoint connection. This will
+// be passed to the server through the listener. The function blocks until the
+// server has finished reading the message.
+func (l *fakeListener) SendRequest(request []byte) error {
+	wrote, err := l.clientEndpoint.Write(request)
+	if requestLen := len(request); wrote != requestLen {
```
Comment (kele) [inline requestparsing/test_harness.go:80]:
```
n, err := l.clientEndpoint.Write(request)
if n != len(request) {
```

`wrote` doesn't carry much meaning, and `requestLen` is not more readable than `len(request)`.

File: requestparsing/test_harness.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+		return errors.New("client connection failed to write the entire request")
+	}
+	return err
+}
+
+// readResponse reads the response from the clientEndpoint connection, sent by
+// the listening server. It will block until the server has sent a response
+// or has timed out.
+func (l *fakeListener) readResponse() ([]byte, error) {
+	// TODO(maramihali@, grenfeldt@): refactor this
```
Comment (kele) [inline requestparsing/test_harness.go:90]: Refactor this how?

File: requestparsing/test_harness.go [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+// server has finished reading the message.
+func (l *fakeListener) SendRequest(request []byte) error {
+	wrote, err := l.clientEndpoint.Write(request)
+	if requestLen := len(request); wrote != requestLen {
+		return errors.New("client connection failed to write the entire request")
+	}
+	return err
+}
+
+// readResponse reads the response from the clientEndpoint connection, sent by
```
Comment (kele) [inline requestparsing/test_harness.go:86]: Mention that you've hardcoded the 4096 limit.

File: requestparsing/test_harness.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+// the listening server. It will block until the server has sent a response
+// or has timed out.
+func (l *fakeListener) readResponse() ([]byte, error) {
```
Comment (kele) [inline requestparsing/test_harness.go:89]: Consider passing a byte slice instead of returning it, similarily to http://go/godoc/bytes/#Reader.Read.

Comment (mattiasgrenfeldt) [reply requestparsing/test_harness.go]: We opted for this solution

File: requestparsing/test_harness.go (hunk 6 above)
Comment (mattiasgrenfeldt) [reply requestparsing/test_harness.go:68]: Refactored in 76e7795

File: requestparsing/test_harness.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattiasgrenfeldt) [reply requestparsing/test_harness.go]: Renamed in 76e7795.

Why should this struct be public (exported)? We will only use it in this package

[the remark this answers is not in the document]
Comment (kele) [reply requestparsing/test_harness.go]: The setup seems to be good enough to be useful for more than just the "requestparsing" package, therefore I thought it might be a good idea to make it public and put into a `testing` package.

File: requestparsing/test_harness.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+// Addr returns the network address of the client endpoint.
+func (l *fakeListener) Addr() net.Addr {
+	return l.clientEndpoint.LocalAddr()
+}
+
+// SendRequest writes a request to the client endpoint connection. This will be passed to the server through the listener.
+// The function blocks until the server has finished reading the message.
+func (l *fakeListener) sendRequest(request []byte) error {
+	n, err := l.clientEndpoint.Write(request)
```
Comment (kele) [inline requestparsing/test_harness.go:87]: Check the error first, then the length.

Usually, if the returned error is not `nil`, you should not assume anything about the other values.

File: requestparsing/test_harness.go [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+// readResponse reads the response from the clientEndpoint connection, sent by the listening server.
+// It will block until the server has sent a response.
+func (l *fakeListener) readResponse(bytes []byte) (int, error) {
+	return l.clientEndpoint.Read(bytes)
+}
+
+// MakeRequest instantiates a new http.Server, sends the request provided as argument and returns the response.
+// 'callback' will be called in the http.Handler with the http.Request that the handler receives.
+// The size of the response is hardcoded as 4096 bytes. If the response received is larger, an error will be returned.
```
Comment (kele) [inline requestparsing/test_harness.go:102]: nit: change "hardcoded" to "limited".

File: requestparsing/test_harness.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kele) [reply requestparsing/test_harness.go]: The main reason is: I imagine this package being quite useful. Keeping it inside a "requestparsing" package makes it visible only to it, therefore no tests in other packages can be made.

File: requestparsing/test_harness.go [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+	if err := listener.sendRequest(req); err != nil {
+		return nil, err
+	}
+
+	resp := make([]byte, 4096)
+	n, err := listener.readResponse(resp)
+	if n == 4096 {
+		return nil, errors.New("response larger than or equal to 4096 bytes")
+	}
+	if err != nil {
```
Comment (kele) [inline requestparsing/test_harness.go:121]: Similarily here, first deal with the error, then consider the rest.

File: requestparsing/test_harness.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mihalimara22) [reply requestparsing/test_harness.go]: Renamed package in [527b6aa](url)

File: requestparsing/test_harness.go (hunk 12 above)
Comment (mihalimara22) [reply requestparsing/test_harness.go]: Resolved in [527b6aa](url)

File: requestparsing/test_harness.go (hunk 13 above)
Comment (mihalimara22) [reply requestparsing/test_harness.go]: Resolved in [527b6aa](url)

File: requestparsing/test_harness.go (hunk 14 above)
Comment (mihalimara22) [reply requestparsing/test_harness.go]: Resolved in [527b6aa](url)
